src/compiler/checker.ts

```diff
@@ -34618,6 +34619,23 @@ namespace ts {
                         return undefined;
                 }
             }
+
+            function checkNaNEquality(errorNode: Node | undefined, operator: SyntaxKind, left: Expression, right: Expression, leftType: Type, rightType: Type) {
+                const isLeftNaN = leftType.flags & TypeFlags.Number && isNaN(skipParentheses(left));
+                const isRightNaN = rightType.flags & TypeFlags.Number && isNaN(skipParentheses(right));
+                if (isLeftNaN || isRightNaN) {
+                    const err = error(errorNode, Diagnostics.This_condition_will_always_return_0,
+                        tokenToString(operator === SyntaxKind.EqualsEqualsEqualsToken || operator === SyntaxKind.EqualsEqualsToken ? SyntaxKind.FalseKeyword : SyntaxKind.TrueKeyword));
+                    if (isLeftNaN && isRightNaN) return;
+                    const location = isLeftNaN ? right : left;
+                    addRelatedInfo(err, createDiagnosticForNode(location, Diagnostics.Did_you_mean_0,
+                        `${operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken ? tokenToString(SyntaxKind.ExclamationToken) : ""}Number.isNaN(${getTextOfNode(location)})`));

```

Comment:
If the expression is too complicated, you might want to swap to using `...`.

So if you can, use

```ts
const location = skipParentheses(isLeftNaN ? right : left);
const locationString = isEntityNameExpression(location) ? entityNameToString(location) : "...";
addRelatedInfo(...);
```

Reply:
What do you think about using the `defaultMaximumTruncationLength`? I think suggestions like Number.isNaN(a[0]) would be better than `Number.isNaN(....)`. Is it better to only allow this suggestion for `EntityNameExpression`s?

```ts
const expression = isLeftNaN ? right : left;
const expressionText = getTextOfNode(skipParentheses(expression), /*includeTrivia*/ false);
addRelatedInfo(err, createDiagnosticForNode(expression, Diagnostics.Did_you_mean_0,
    `${operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken
        ? tokenToString(SyntaxKind.ExclamationToken)
        : ""}Number.isNaN(${expressionText.length > defaultMaximumTruncationLength ? "..." : expressionText})`));
```

Reply:
I think it's fine to just limit it to `EntityNameExpression`s. The problem is not necessarily just about length, it's also about multi-line code and indentation too.